src/msw/joystick.cpp

```diff
@@ -37,8 +37,129 @@
 
 #include <regstr.h>
 
+// Use optimised count trailing zeros where available.
+#ifdef __GNUC__
+    #define ctz(x) __builtin_ctz(x)
+#else
+int ctz(unsigned x)
+{
+   int n;
+
+   if (x == 0) return(32);
+   n = 1;
+   if ((x & 0x0000FFFF) == 0) {n = n +16; x = x >>16;}
+   if ((x & 0x000000FF) == 0) {n = n + 8; x = x >> 8;}
+   if ((x & 0x0000000F) == 0) {n = n + 4; x = x >> 4;}
+   if ((x & 0x00000003) == 0) {n = n + 2; x = x >> 2;}
+   return n - (x & 1);
+}
+#endif
+
+enum {
+    wxJS_AXIS_X = 0,
+    wxJS_AXIS_Y,
+    wxJS_AXIS_Z,
+    wxJS_AXIS_RUDDER,
+    wxJS_AXIS_U,
+    wxJS_AXIS_V,
+
+    wxJS_AXIS_MAX = 32767,
+    wxJS_AXIS_MIN = -32767,
+    wxJS_MAX_AXES = 6,      // WinMM supports up to 6 axes.
+    wxJS_MAX_BUTTONS = 32,  // WinMM supports up to 32 buttons.
+};
+
 wxIMPLEMENT_DYNAMIC_CLASS(wxJoystick, wxObject);
 
+////////////////////////////////////////////////////////////////////////////
+// Background thread for reading the joystick device
+////////////////////////////////////////////////////////////////////////////
+
+class wxJoystickThread : public wxThread
+{
+public:
+    explicit wxJoystickThread(int joystick);
+    void* Entry() wxOVERRIDE;
+    void SetPolling(wxWindow* win, int pollingFreq)
+    {
+        m_catchwin = win;
+        m_polling = pollingFreq;
+    };
+
+
+private:
+    void      SendEvent(wxEventType type, long ts, int change = 0);
+    int       m_joystick;
+    int       m_buttons;
+    wxWindow* m_catchwin;
+    int       m_polling;
+    JOYINFO   m_joyInfo;
+    JOYINFO   m_lastJoyInfo;
+};
+
+
+wxJoystickThread::wxJoystickThread(int joystick)
+    : m_joystick(joystick),
+      m_buttons(0),
+      m_catchwin(NULL),
+      m_polling(0),
+      m_joyInfo(),
+      m_lastJoyInfo()
+{
+}
+
+void wxJoystickThread::SendEvent(wxEventType type, long ts, int change)
+{
+    wxJoystickEvent joystickEvent(type, m_buttons, m_joystick, change);
+
+    joystickEvent.SetTimestamp(ts);
+    joystickEvent.SetPosition(wxPoint(m_joyInfo.wXpos, m_joyInfo.wYpos));
+    joystickEvent.SetZPosition(m_joyInfo.wZpos);
+    joystickEvent.SetEventObject(m_catchwin);
+
+    if (m_catchwin)
+        m_catchwin->GetEventHandler()->ProcessThreadEvent(joystickEvent);
+}
+
+void* wxJoystickThread::Entry()
+{
+    joyGetPos(m_joystick, &m_lastJoyInfo);
+
+    while (!TestDestroy())
+    {
+        Sleep(m_polling);
+        DWORD ts = GetTickCount();
+
+        joyGetPos(m_joystick, &m_joyInfo);
+        m_buttons = m_joyInfo.wButtons;
+        DWORD delta = m_buttons ^ m_lastJoyInfo.wButtons;
+        DWORD deltaUp = delta & !m_buttons;
+        DWORD deltaDown = delta & m_buttons;
```

Comment:
Note that you use 3 different types interchanged: `int` for `m_buttons`, `DWORD` (= `unsigned long`) for local variables, and `UINT` (= `unsigned int`) for the `m_joyInfo` members. I suggest using `unsigned int` everywhere.

Reply:
Good catch - thanks. The DWORDS are because I originally used joyGetPosEx and JOYINFOEX (with DWORD members), but then I noticed this isn't supported by the Borland compiler, so dropped back to joyGetPos and JOYINFO (with UINT members).

Reply:
The ```JOYINFO``` members are declared ```UINT``` in the external header (```mmsystem.h```) for both MS and GNU compilers. Would you still suggest ```unsigned int``` in preference to ```UINT```?

Reply:
They are the same (at least on my Windows SDK): `typedef unsigned int        UINT;`. I don't  know if wxWidgets has some guidelines over which to prefer.
Just make sure to they are the same type when performing the bit operations to prevent signedness and conversion warnings.

Also make sure to cast to the correct type when using them later on. For example `wxJoystickEvent` expects `ts` to be of type `long` and `m_buttons` of type `int`.  And `SetPosition(wxPoint)` and  `SetZPosition` also use `int`.

Reply:
Thanks. Hopefully fixed in c4a894ddc2e3821072f4717b1f21f09c758a5bd8.
I've gone with ```UINT`` for m_buttons and the derived variables, because that's what the external functions that write to them expect. and using the same for derived variables eliminates implicit casts. I've added explicit casts where appropriate.